soc/riscv/litex-vexriscv/Kconfig.soc

```diff
@@ -9,6 +9,7 @@ config SOC_RISCV32_LITEX_VEXRISCV
 	select RISCV_ISA_RV32I
 	select RISCV_ISA_EXT_M
 	select RISCV_ISA_EXT_A
+	select ARCH_SUPPORTS_CODE_DATA_RELOCATION
```

Comment:
ok, I found this a bit confusing because you are setting an `ARCH_*` property from a SoC configuration file. The `ARCH_HAS / ARCH_SUPPORTS` should only be enabled by the architecture (like the name is suggesting).

`ARCH_HAS_CODE_DATA_RELOCATION` should be enable at arch level in `arch/riscv/Kconfig` (because you indeed modified `include/zephyr/arch/riscv/common/linker.ld`) while at the SoC level you must only set `CONFIG_CODE_DATA_RELOCATION` for the SoCs that actually supports it.

Reply:
If `ARCH_HAS_CODE_DATA_RELOCATION` is selected at top-level `riscv`, then we still need to express whether a given SoC actually uses `arch/riscv/common/linker.ld`. It seems like we'd need to add a `riscv`-specific `SOC_HAS_CODE_DATA_RELOCATION` symbol that is selected by those SoCs that support it, since `CODE_DATA_RELOCATION` should be prevented from being enabled unless the system actually supports it. I don't mind adding that layer of indirection, but it seems a bit unnecessary only to avoid setting an `ARCH_` symbol at SoC-level (though I grant that's a pretty minor stylistic opinion).

Unless you had something else in mind?

Reply:
I missed that some SoCs are not actually using `arch/riscv/common/linker.ld` (why thought?), this sucks.

Can you point me to the SoC not using the common linker script?

Honestly I think that the error is not using the commen-arch linker script here, so my suggestion here is to proceed to fix this using the `ARCH_HAS_` in `arch/riscv/Kconfig`. @stephanosio any better idea here?

Reply:
> Can you point me to the SoC not using the common linker script?

The major SoC that doesn't use the common one is `esp32c3` (it has a lot of unusual things in its linker script) but there were a few others I don't recall offhand.

Reply:
Ok, in general this is a SoC limitation. From the ARCH prospective nothing is preventing the usage of the code relocation. So IMO using `ARCH_HAS_CODE_DATA_RELOCATION` for all the RISCV boards is correct. `esp32c3` and others should simply not set `CODE_DATA_RELOCATION`.

Reply:
Alright; changed the top-level `RISCV` config to select `ARCH_HAS_CODE_DATA_RELOCATOIN`.